File: 1_twosum.py

def two_sum(nums, target):
    rests = {}
    for n in nums:
        if n in rests.keys():
            return (rests[n], n)
        rest = target - n
        if rest not in rests:
            rests[rest] = n
    return -1

nums = [2, 7, 1, 12, 49, 11, 15]
targets = (9, 17, 11, 56)
for target in targets:
    twosum = two_sum(nums, target)
    print '%d is %d + %d' % (target, twosum[0], twosum[1]) if twosum != -1 else '%d is not achievable' % target

File: 1_twosum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;


class Solution {
    public:
        vector<int> twoSum(vector<int>& nums, int target) {
            vector<int> res;
            unordered_map<int, int> rests;
            for(const int& num:nums) {
                int rest = target - num;
                if(rests.find(num) != rests.end()) {
                    res.push_back(rests[num]);
                    res.push_back(num);
                    return res;
                }
                rests[rest] = num;
            }
        }
};


int main() {
    Solution s;
    vector<int> nums = {2, 7, 11, 15};
    vector<int> sol;
    sol = s.twoSum(nums, 9);
    for(const int& n:sol) {
        cout << n << " ";
    }
    cout << endl;
    return 0;
}

